flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorTest.java

```diff
@@ -1662,6 +1665,82 @@ public void testTerminationOnFatalError() throws Throwable {
 		}
 	}
 
+	/**
+	 * Tests that task manager will sync slot with job master throw heartbeat (See FLINK-11059).
+	 */
+	@Test(timeout = 10000L)
+	public void testSyncSlotsWithJobMasterByHeartbeat() throws Exception {
+		final TaskSlotTable taskSlotTable = new TaskSlotTable(
+				Arrays.asList(ResourceProfile.UNKNOWN, ResourceProfile.UNKNOWN),
+				timerService);
+		final TaskManagerServices taskManagerServices = new TaskManagerServicesBuilder().setTaskSlotTable(taskSlotTable).build();
+
+		final TaskExecutor taskExecutor = createTaskExecutor(taskManagerServices);
+
+		final TestingResourceManagerGateway testingResourceManagerGateway = new TestingResourceManagerGateway();
+
+		final CompletableFuture<AllocationID> allocationNotifiedFree = new CompletableFuture<>();
+
+		final CompletableFuture<Void> initialSlotReportFuture = new CompletableFuture<>();
+		testingResourceManagerGateway.setSendSlotReportFunction(resourceIDInstanceIDSlotReportTuple3 -> {
+			initialSlotReportFuture.complete(null);
+			return CompletableFuture.completedFuture(Acknowledge.get());
+
+		});
+
+		testingResourceManagerGateway.setNotifySlotAvailableConsumer(instanceIDSlotIDAllocationIDTuple3 ->
+			allocationNotifiedFree.complete(instanceIDSlotIDAllocationIDTuple3.f2));
```

Comment:
Instead of using a `CompletableFuture` which can only be completed once we could use a `ArrayBlockingQueue` with a capacity of `2`. That way we could assert (probabilistically) via a `queue.poll(5L, TimeUnit.MILLISECONDS)` that no other slots are freed. That way we could guard against future changes which could make this test pass if always the `allocationIdOnlyInTM` is freed first if multiple slots are being freed.

Reply:
OK, make sense.